packages/block-editor/src/components/embedded-admin-context/index.js

```diff
@@ -0,0 +1,103 @@
+/**
+ * WordPress dependencies
+ */
+import { useRefEffect } from '@wordpress/compose';
+import { useState, createPortal } from '@wordpress/element';
+import { ENTER, SPACE, ESCAPE } from '@wordpress/keycodes';
+import { focus } from '@wordpress/dom';
+import { __experimentalStyleProvider as StyleProvider } from '@wordpress/components';
+
+export default function EmbeddedAdminContext( props ) {
+	const [ shadow, setShadow ] = useState();
+	const [ hasFocus, setHasFocus ] = useState();
+	const ref = useRefEffect( ( element ) => {
+		const root = element.attachShadow( { mode: 'open' } );
+		const style = document.createElement( 'style' );
+		Array.from( document.styleSheets ).forEach( ( styleSheet ) => {
+			if ( styleSheet.ownerNode.getAttribute( 'data-emotion' ) ) {
+				return;
+			}
+
+			// Try to avoid requests for stylesheets of which we already
+			// know the CSS rules.
+			try {
+				let cssText = '';
+
+				for ( const cssRule of styleSheet.cssRules ) {
+					cssText += cssRule.cssText;
+				}
+
+				style.textContent += cssText;
+			} catch ( e ) {
+				root.appendChild( styleSheet.ownerNode.cloneNode( true ) );
+			}
+		} );
+		root.appendChild( style );
+		setShadow( root );
+
+		function onFocusIn() {
+			setHasFocus( true );
+		}
+
+		function onFocusOut() {
+			setHasFocus( false );
+		}
+
+		function onKeyDown( event ) {
+			if ( element !== event.path[ 0 ] ) return;
+			if ( event.keyCode !== ENTER && event.keyCode !== SPACE ) return;
+
+			focus.tabbable.find( root )[ 0 ].focus();
+			event.preventDefault();
+		}
+
+		function onRootKeyDown( event ) {
+			if ( event.keyCode !== ESCAPE ) return;
+
+			root.host.focus();
+			event.preventDefault();
+		}
+
+		let timeoutId;
+
+		// Allow clicking through the placeholder so focus moves to the block
+		// wraper, which can handle delete, enter, etc.
+		function onMouseDown() {
+			element.removeAttribute( 'tabindex' );
+			timeoutId = setTimeout( () =>
+				element.setAttribute( 'tabindex', '0' )
+			);
+		}
+
+		root.addEventListener( 'focusin', onFocusIn );
+		root.addEventListener( 'focusout', onFocusOut );
+		root.addEventListener( 'keydown', onRootKeyDown );
+		element.addEventListener( 'keydown', onKeyDown );
+		element.addEventListener( 'mousedown', onMouseDown );
+		return () => {
+			root.removeEventListener( 'focusin', onFocusIn );
+			root.removeEventListener( 'focusout', onFocusOut );
+			root.removeEventListener( 'keydown', onRootKeyDown );
+			element.removeEventListener( 'keydown', onKeyDown );
+			element.removeEventListener( 'mousedown', onMouseDown );
+			clearTimeout( timeoutId );
+		};
+	}, [] );
```

Comment:
Can you explain a bit more the logic in this component. I'm unsure what all of this is about.

Reply:
Added some comments.